clients/src/main/java/org/apache/kafka/clients/admin/internals/AdminApiDriver.java

```diff
@@ -0,0 +1,476 @@
+        List<RequestSpec<K>> requests,
+        BiMultimap<T, K> multimap,
+        BiFunction<Set<K>, T, AbstractRequest.Builder<?>> buildRequest
+    ) {
+        for (Map.Entry<T, Set<K>> entry : multimap.entrySet()) {
+            T scope = entry.getKey();
+
+            Set<K> keys = entry.getValue();
+            if (keys.isEmpty()) {
+                continue;
+            }
+
+            RequestState requestState = requestStates.computeIfAbsent(scope, c -> new RequestState());
+            if (requestState.hasInflight()) {
+                continue;
+            }
+
+            AbstractRequest.Builder<?> request = buildRequest.apply(keys, scope);
+            RequestSpec<K> spec = new RequestSpec<>(
+                handler.apiName() + "(api=" + request.apiKey() + ")",
+                scope,
+                new HashSet<>(keys), // copy to avoid exposing mutable state
+                request,
+                requestState.nextAllowedRetryMs,
+                deadlineMs,
+                requestState.tries
+            );
+
+            requestState.setInflight(spec);
+            requests.add(spec);
+        }
+    }
+
+    private void collectLookupRequests(List<RequestSpec<K>> requests) {
+        dynamicMapping.ifPresent(mapping -> {
+            collectRequests(
+                requests,
+                lookupMap,
+                (keys, scope) -> mapping.lookupStrategy.buildRequest(keys)
+            );
+        });
+    }
+
+    private void collectFulfillmentRequests(List<RequestSpec<K>> requests) {
+        collectRequests(
+            requests,
+            fulfillmentMap,
+            (keys, scope) -> handler.buildRequest(scope.destinationBrokerId, keys)
+        );
+    }
+
+    /**
+     * This is a helper class which helps us to map requests that need to be sent
+     * to the internal `Call` implementation that is used internally in
+     * {@link org.apache.kafka.clients.admin.KafkaAdminClient}.
+     */
+    public static class RequestSpec<K> {
+        public final String name;
+        public final RequestScope scope;
+        public final Set<K> keys;
+        public final AbstractRequest.Builder<?> request;
+        public final long nextAllowedTryMs;
+        public final long deadlineMs;
+        public final int tries;
+
+        public RequestSpec(
+            String name,
+            RequestScope scope,
+            Set<K> keys,
+            AbstractRequest.Builder<?> request,
+            long nextAllowedTryMs,
+            long deadlineMs,
+            int tries
+        ) {
+            this.name = name;
+            this.scope = scope;
+            this.keys = keys;
+            this.request = request;
+            this.nextAllowedTryMs = nextAllowedTryMs;
+            this.deadlineMs = deadlineMs;
+            this.tries = tries;
+        }
+
+        @Override
+        public String toString() {
+            return "RequestSpec(" +
+                "name='" + name + '\'' +
```

Comment:
Is the `'\'` intentional here? It looks weird.

Reply:
Looks like I was trying to quote the name. It's probably overkill.